pkg/controller/acmechallenges/sync.go

```diff
@@ -88,6 +88,14 @@ func (c *controller) Sync(ctx context.Context, ch *cmacme.Challenge) (err error)
 		return nil
 	}
 
+	// This finalizer ensures that the challenge is not garbage collected before
+	// cert-manager has a chance to clean up resources created for the
+	// challenge.
+	if finalizerRequired(ch) {
+		ch, err = addFinalizer(c, ctx, ch)
+		return err
```

Comment:
> Suggestion: do you think it might make sense to move the actual applying to cluster to the deferred function (as that's how we do it elsew[here](https://github.com/cert-manager/cert-manager/blob/master/pkg/controller/certificaterequests/sync.go#L57), i.e certificate-requests controllers here where annotations are applied in the deferred function.

I do think that's a good idea, and it would be nice to be consistent.
But with added unit-tests I think will bloat this PR too much. 
Let me merge this and try that approach in a followup PR. 

> (I was also wondering whether we might want to not return here, but just carry on with the next step. However, perhaps that would introduce a risk of creating some resources in cluster that don't get cleaned up if applying of the finalizer fails.)

I prefer to stick to the formula that the Sync function performs a sequence of steps but only ever performs one of these per call. It makes it easier to test it.

Reply:
Sounds good and makes sense to me